src/Microsoft.AspNet.Mvc.Core/UnitTestability/MvcControllerUnitTestHelper.cs

```diff
@@ -0,0 +1,277 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Net.WebSockets;
+using System.Security.Claims;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Http.Core;
+using Microsoft.AspNet.Http.Core.Collections;
+using Microsoft.AspNet.Http.Security;
+using Microsoft.AspNet.Routing;
+using Microsoft.Framework.DependencyInjection;
+using Microsoft.Framework.DependencyInjection.Fallback;
+
+// Not sure if this is the right location/namespace for this..
+namespace Microsoft.AspNet.Mvc.Core
+{
+	public class MvcControllerUnitTestHelper
+	{
+		public MvcControllerUnitTestHelper() : this(new ServiceCollection(), new RouteData())
+		{
+		}
+
+		public MvcControllerUnitTestHelper(IServiceCollection serviceCollection)
+			: this(serviceCollection, new RouteData())
+		{
+		}
+
+		private MvcControllerUnitTestHelper(IServiceCollection serviceCollection, RouteData routeData)
+		{
+			RouteData = routeData;
+			ServiceProvider = serviceCollection.BuildServiceProvider();
+		}
+
+		// Not sure if this RouteData property will be usefull..
+		public RouteData RouteData
+		{
+			get;
+			set;
+		}
+
+		public IServiceProvider ServiceProvider
+		{
+			get;
+			set;
+		}
+
+		// Not sure if this HttpContext property will be usefull..
+		public HttpContext HttpContext
+		{
+			get;
+			set;
+		}
+
+		public Controller Initialize(Controller controller, MvcControllerUnitTestHelperCallback callback)
+		{
+			return Initialize(controller, new ActionDescriptor(), callback);
+			// Question: Can we build (fully or partially) ActionDescriptor() from tyep information??
+		}
+
+		private Controller Initialize(Controller controller, ActionDescriptor actionDescriptor, MvcControllerUnitTestHelperCallback callback)
+		{
+			HttpContext = new MvcControllerTestHelperHttpContext();
+			HttpContext.RequestServices = ServiceProvider;
+			controller.ActionContext = new ActionContext(HttpContext, RouteData, actionDescriptor);
+			controller.Url = new MvcControllerTestHelpeUrlHelper();
+
+			if (callback != null)
+			{
+				((MvcControllerTestHelperHttpContext)HttpContext).OnRequestAborted = callback.OnRquestAborted;
+				((MvcControllerTestHelperHttpContext)HttpContext).OnRequestFormCollection = callback.OnRequestFormCollection;
+				((MvcControllerTestHelpeUrlHelper)controller.Url).OnAction =
+						(action, controllerName, values, protocol, host, fragment) => callback.OnUrlAction(action);
+			}
+
+			return controller;
+		}
+
+		private class MvcControllerTestHelperHttpContext : HttpContext
```

Comment:
We should use `DefaultHttpContext` here. My guess is that we want any testing helper code to go into a separate project and not `Mvc.Core`. Then it's no trouble to introduce the dependency on the package that `DefaultHttpContext` lives in.